api/librdb-api.h

```diff
@@ -155,26 +174,42 @@ typedef struct RdbHandlersRawCallbacks {
     RdbRes (*handleBegin)(RdbParser *p, void *userData, size_t size);
     RdbRes (*handleFrag)(RdbParser *p, void *userData, RdbBulk frag);
     RdbRes (*handleEnd)(RdbParser *p, void *userData);
-//    RdbRes (*handleBeginModuleAux)(RdbParser *p, void *userData, RdbBulk name, int encver, int when);
+
+    /*** TODO: RdbHandlersRawCallbacks: handleBeginModuleAux ***/
+    RdbRes (*handleBeginModuleAux)(RdbParser *p, void *userData, RdbBulk name, int encver, int when);
+
 } RdbHandlersRawCallbacks;
 
 typedef struct RdbHandlersStructCallbacks {
     HANDLERS_COMMON_CALLBACKS
     RdbRes (*handleStringValue)(RdbParser *p, void *userData, RdbBulk str);
-    RdbRes (*handlerQListNode)(RdbParser *p, void *userData, RdbBulk listNode);
-    RdbRes (*handlerPlainNode)(RdbParser *p, void *userData, RdbBulk node);
-    /* ... TBD ... */
+    RdbRes (*handleQListNode)(RdbParser *p, void *userData, RdbBulk listNode);
+    RdbRes (*handlePlainNode)(RdbParser *p, void *userData, RdbBulk node);
+
+    /*** TODO: RdbHandlersStructCallbacks: handlerHashListPack, handleSet, handleZsetListPack, handleFunction, handleStreamListPack, handleStreamMetadata, handleStreamCgroup ***/
+    RdbRes (*handlerHashListPack)(RdbParser *p, void *userData, RdbBulk hash);
+    RdbRes (*handleSet)(RdbParser *p, void *userData, RdbBulk intSet);
+    RdbRes (*handleZsetListPack)(RdbParser *p, void *userData, RdbBulk zset);
+    RdbRes (*handleFunction)(RdbParser *p, void *userData, RdbBulk func);
+    RdbRes (*handleStreamListPack)(RdbParser *p, void *userData, RdbBulk nodekey, RdbBulk listpack);
+    RdbRes (*handleStreamMetadata)(RdbParser *p, void *userData, uint64_t len, uint64_t entriesAdded, streamID *firstID, streamID *lastID, streamID *maxDeletedEntryID);
+    RdbRes (*handleStreamCgroup)(RdbParser *p, void *userData, streamCG *cgroup); /* when succeeds, 'cgroup' belongs to the calee  */
```

Comment:
I think we should use `streamCG` because:
1. It is more complex collection and a set of callbacks might won't be clear enough in this case.
2. If requested integrity-check then we will need to allocate internally streamCG anyway (assuming that the integrity-check will happen in the parser and not by app or by callback extensions. Otherwise it is more complicated, especially in the case of RESTORE).
3. If in the future Redis will be linked to librdb, then the flow will be already optimized.

WDYT?

Reply:
i don't think we wanna bind redis to use the same struct that's used by the parser.
same as we don't bind them to use the same robj, or sds, and instead we let the app get raw callbacks and store things as it wishes.
in the case of the general stream-level metadata, a struct could be ok.
if we want ABI compatibility, so the app doesn't have to run with the same parser version it compiled with, then we'll also need some version field, so the caller can know which fields are valid to read from.

i don't think the parser should do integrity checks, certainly not high level / logical ones (maybe just low level format ones), that should be the app's responsibility.

Reply:
merging the threads about stream metadata and cgroups into one...

i think that the same as we break a listpack into user friendly records in the high level CBs, we should also break up the consumer PELs, rather than hand the user a rax.
let's think about someone that wants to either generate a JSON, or send a list of XGROUP, XCLAIM commands (like what aof.c does).

regarding the struct CBs, i'm not sure if any of that is really useful for memory estimation, redis itself can't easily estimate the memory of a rax tree.
the other use case is if redis would use the parser to create it's internal state, but here, unlike listpack, redis doesn't store the data from the rdb directly into RAM... in any case it re-constructs a rax when parsing the file, so it can listen for the high level callbacks and do that from there.

Reply:
Ok. So as i see it, we can supply yet another set of callback handlers that will do integ-check. 

It will be little cumbersome to handle RESTORE payload - maybe we should pass hints to callbacks.

Reply:
this may mean that for the concept of multi-level callback binding, we might wanna let an application use one level for the stream records, and another one for the consumer groups.

Reply:
@oranagra , I changed the API callbacks to be iterative (SAX) rather than providing a stream data-structure (DOM). Please review.